### PR TITLE
Add Evidences to Storage

### LINKED ISSUE #649: Support persistence in evidence store
**Is your feature request related to a problem? Please describe.**

Currently, the evidences in the evidence store are only stored read-only.

**Describe the solution you'd like**

We should use the persistence layer.

### LINKED ISSUE #833: Add missing authorisation checks to all services
```[tasklist]
- [ ] https://github.com/clouditor/clouditor/issues/1060
- [x] `ListEvidences` in Evidence Store service? (certification stuff, e.g. ListCertification)
- [ ] #1056
- [ ] https://github.com/clouditor/clouditor/issues/1058
- [ ] https://github.com/clouditor/clouditor/issues/1059
```

_Originally posted by @lebogg in https://github.com/clouditor/clouditor/pull/824#pullrequestreview-1156198326_

### PR DESCRIPTION
**Pull request by lebogg**
Resolves #649

This also adds authorisation checks to the evidence store and partially resolves #833

### CODE REVIEW

File: service/evidence/evidence_store.go [hunk 1]
```diff
@@ -43,60 +46,109 @@ var log *logrus.Entry
...
+	// Check, if this request has access to the cloud service according to our authorization strategy.
+	if !svc.authz.CheckAccess(ctx, service.AccessUpdate, req) {
+		return nil, service.ErrPermissionDenied
 	}
 
-	s.evidences[req.Evidence.Id] = req.Evidence
-	go s.informHooks(req.Evidence, nil)
+	err = svc.storage.Create(req.Evidence)
+	if err != nil {
+		// TODO(lebogg): Almost identical error handling as above -> extract here or even to internal
```
Comment (oxisto) [inline service/evidence/evidence_store.go:125]: This code here by @lebogg might need to be adjusted to the "new" behaviour of our streaming functions such as AssessEvidence.

File: persistence/gorm/serializer.go [hunk 2]
```diff
@@ -132,3 +133,48 @@ func (AnySerializer) Scan(ctx context.Context, field *schema.Field, dst reflect.
 	field.ReflectValueOf(ctx, dst).Set(reflect.ValueOf(&a))
 	return
 }
+
+// StructpbValueSerializer is a GORM serializer that allows the serialization and deserialization of the
+// google.protobuf.Value protobuf message type.
+type StructpbValueSerializer struct{}
```
Comment (oxisto) [inline persistence/gorm/serializer.go:139]: Should probably be just named ValueSerializer

File: persistence/gorm/serializer.go [hunk 3]
```diff
@@ -132,3 +133,48 @@ func (AnySerializer) Scan(ctx context.Context, field *schema.Field, dst reflect.
...
+
+// Value implements https://pkg.go.dev/gorm.io/gorm/schema#SerializerValuerInterface to indicate
+// how this struct will be saved into an SQL database field.
+func (StructpbValueSerializer) Value(_ context.Context, _ *schema.Field, _ reflect.Value, fieldValue interface{}) (interface{}, error) {
+	var (
+		v  *structpb.Value
+		ok bool
+	)
+
+	if fieldValue == nil {
```
Comment (oxisto) [inline persistence/gorm/serializer.go:149]: We had some issues with "nil" in other serialisers, we need to double check if the nil checks are the same in every serializer

File: service/evidence/evidence_store.go [hunk 4]
```diff
@@ -43,50 +46,88 @@ var log *logrus.Entry
...
 
 // StoreEvidence is a method implementation of the evidenceServer interface: It receives a req and stores it
-func (s *Service) StoreEvidence(_ context.Context, req *evidence.StoreEvidenceRequest) (resp *evidence.StoreEvidenceResponse, err error) {
-	resp = &evidence.StoreEvidenceResponse{}
-
+func (svc *Service) StoreEvidence(ctx context.Context, req *evidence.StoreEvidenceRequest) (res *evidence.StoreEvidenceResponse, err error) {
 	// Validate request
 	err = service.ValidateRequest(req)
 	if err != nil {
 		log.Error(err)
```
Comment (anatheka) [inline service/evidence/evidence_store.go:104]:
```suggestion
```

File: service/evidence/evidence_store_test.go [hunk 5]
```diff
@@ -80,15 +107,15 @@ func TestService_StoreEvidence(t *testing.T) {
...
 	}{
 		{
 			name: "Store req to the map",
 			args: args{
 				in0: context.TODO(),
 				req: &evidence.StoreEvidenceRequest{
 					Evidence: &evidence.Evidence{
-						Id:             "11111111-1111-1111-1111-111111111111",
+						Id:             testdata.MockEvidenceID,
 						CloudServiceId: "00000000-0000-0000-0000-000000000000",
```
Comment (anatheka) [inline service/evidence/evidence_store_test.go:119]:
```suggestion
						CloudServiceId: testdata.MockCloudServiceID,
```

File: service/evidence/evidence_store_test.go [hunk 6]
```diff
@@ -80,15 +107,15 @@ func TestService_StoreEvidence(t *testing.T) {
...
 						ToolId:         "MockTool",
```
Comment (anatheka) [inline service/evidence/evidence_store_test.go:120]:
```suggestion
						ToolId:         testutil.MockEvidenceToolID,
```

File: service/evidence/evidence_store_test.go [hunk 7]
```diff
@@ -109,7 +136,7 @@ func TestService_StoreEvidence(t *testing.T) {
 				in0: context.TODO(),
 				req: &evidence.StoreEvidenceRequest{
 					Evidence: &evidence.Evidence{
-						Id:             "11111111-1111-1111-1111-111111111111",
+						Id:             testdata.MockEvidenceID,
 						CloudServiceId: "00000000-0000-0000-0000-000000000000",
```
Comment (anatheka) [inline service/evidence/evidence_store_test.go:140]:
```suggestion
						CloudServiceId: testdata.MockCloudServiceID,
```

**Review by anatheka — COMMENTED** [review]
Only some minor things. However, I don't really know much about the gorm stuff.

File: persistence/gorm/serializer.go [hunk 8]
```diff
@@ -132,3 +133,48 @@ func (AnySerializer) Scan(ctx context.Context, field *schema.Field, dst reflect.
...
 	return
 }
+
+// StructpbValueSerializer is a GORM serializer that allows the serialization and deserialization of the
+// google.protobuf.Value protobuf message type.
+type StructpbValueSerializer struct{}
+
+// Value implements https://pkg.go.dev/gorm.io/gorm/schema#SerializerValuerInterface to indicate
+// how this struct will be saved into an SQL database field.
+func (StructpbValueSerializer) Value(_ context.Context, _ *schema.Field, _ reflect.Value, fieldValue interface{}) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (oxisto) [reply persistence/gorm/serializer.go:143]: The gorm example for a serialiser also implemented it for the non-pointer types (see https://gorm.io/docs/serializer.html#Register-Serializer). I basically followed the exampled there.

File: persistence/gorm/serializer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (oxisto) [reply persistence/gorm/serializer.go:139]: Changed

File: persistence/gorm/serializer.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (anatheka) [reply persistence/gorm/serializer.go]: Ok.

**Review by anatheka — APPROVED** [review]

**Pull request merged**